Updating news and collections

danieltruong opened the pull request:
https://bcmines.atlassian.net/browse/NRPT-333
https://bcmines.atlassian.net/browse/FLCN-37
https://bcmines.atlassian.net/browse/NRPT-367
https://bcmines.atlassian.net/browse/NRPT-368
https://bcmines.atlassian.net/browse/NRPT-377
https://bcmines.atlassian.net/browse/NRPT-388

dhlevi reviewed (changes requested):
That's a lot of work! Code-wise everything looks fine, just a couple of small comments.

Currently though I cannot create/edit collections or news items smoothly. Every attempt to save throws errors. Sometimes when I create a new collection, an error is thrown but the collection is created, sometimes not. Edits seem to always fail. 

Creating news items always throws an error, but seems to create the item successfully. Changes to the Title are ignored on create, but can be changed on update (description/url seems fine).

I'm going to debug a bit and see if I can spot the causes of some of the wonkiness. There are a bunch of tickets related to collection creating and updating as well so not sure entirely what should be covered by all your work here or by those tickets specifically... probably a bunch of overlap!

dhlevi commented on angular/projects/admin-nrpti/src/app/news/news-add-edit/news-add-edit.component.ts:
```diff
@@ -128,19 +128,20 @@ export class NewsAddEditComponent implements OnInit, OnDestroy {
       // Update the news item.
-      this.factoryService.editNews(newsItem).subscribe(async res => {
+      const res = await this.factoryService.editNews(newsItem);
+      if (!res || res.modifiedCount !== 1) {
```
should be:
`if (!res || res.nModified !== 1) {`

or alternatively res.ok !== 1

danieltruong replied:
3a32d6f8c9e96a7ef201e56a13995a6357f2498e

dhlevi commented on angular/projects/admin-nrpti/src/app/news/news-add-edit/news-add-edit.component.ts:
```diff
@@ -128,19 +128,20 @@ export class NewsAddEditComponent implements OnInit, OnDestroy {
-        }
-      });
+      const res = await this.factoryService.createNews(newsItem);
+      if (!res || res.insertedCount !== 1) {
```
Should be:
`if (!res || res.ok !== 1) {`

Or, if you're sending back a response from the API, double check you're sending back the custom response and not the mongo response.

danieltruong replied:
3a32d6f8c9e96a7ef201e56a13995a6357f2498e

dhlevi commented on angular/projects/admin-nrpti/src/app/mines/mines-collections-add-edit/mines-collections-add-edit.component.ts:
```diff
@@ -405,34 +378,30 @@ export class MinesCollectionsAddEditComponent implements OnInit, OnDestroy {
-
-        const createdCollection = res && res.length && res[0] && res[0].length && res[0][0] && res[0][0].object;
+      const res = await this.factoryService.createCollection(collection);
+      if (!res || res.insertedCount !== 1) {
```
Should be: 
`if (!res || res.ok !== 1)`

danieltruong replied:
3a32d6f8c9e96a7ef201e56a13995a6357f2498e

dhlevi commented on angular/projects/admin-nrpti/src/app/mines/mines-collections-add-edit/mines-collections-add-edit.component.ts:
```diff
@@ -405,34 +378,30 @@ export class MinesCollectionsAddEditComponent implements OnInit, OnDestroy {
-        this.updateRecordsCollectionId(this.collection._id);
-
+      const res = await this.factoryService.editCollection(collection);
+      if (!res || res.modifiedCount !== 1) {
```
Should be: 
`if (!res || res.nModified !== 1)` or `if (!res || res.ok !== 1)`

danieltruong replied:
3a32d6f8c9e96a7ef201e56a13995a6357f2498e

dhlevi commented on api/src/controllers/news-controller.js:
```diff
@@ -40,47 +44,79 @@ exports.protectedPut = async function (args, res, next) {
+  incomingObj.description && (news.description = incomingObj.description);
+  incomingObj.projectName && (news.projectName = incomingObj.projectName);
+  incomingObj.type && (news.type = incomingObj.type);
+  incomingObj.type && (news.type = incomingObj.type);
```
Looks like a duped line. I'm guessing it should be:
`incomingObj.title && (news.type = incomingObj.title);`

danieltruong replied:
3a32d6f8c9e96a7ef201e56a13995a6357f2498e

danieltruong commented on api/src/controllers/news-controller.js:
```diff
@@ -40,47 +44,79 @@ exports.protectedPut = async function (args, res, next) {
     defaultLog.info(`protectedPut - error updating news: ${newsId}`);
     defaultLog.debug(error);
-    return queryActions.sendResponse(res, 400, {});
+    console.log(news);
```
3a32d6f8c9e96a7ef201e56a13995a6357f2498e

danieltruong commented on api/src/controllers/record-controller.js:
```diff
@@ -264,6 +259,7 @@ exports.protectedDelete = async function (args, res, next) {
   } catch (error) {
     defaultLog.info(`protectedDelete - error deleting record: ${recordId}`);
     defaultLog.debug(error);
+    console.log(error);
```
3a32d6f8c9e96a7ef201e56a13995a6357f2498e

danieltruong commented on api/src/controllers/record-controller.js:
```diff
@@ -248,6 +242,7 @@ exports.protectedDelete = async function (args, res, next) {
   }
 
   try {
+    console.log(RecordTypeEnum.BCMI_SCHEMA_NAMES);
```
3a32d6f8c9e96a7ef201e56a13995a6357f2498e

danieltruong commented on angular/projects/admin-nrpti/src/app/services/news.service.ts:
```diff
@@ -5,7 +5,28 @@ import { ApiService } from './api.service';
+  }
+
+  public editNews(news: any): Promise<any> {
+    if (!news) {
```
3a32d6f8c9e96a7ef201e56a13995a6357f2498e

mtCarto commented on api/src/controllers/collection-controller.js:
```diff
@@ -106,4 +252,29 @@ exports.protectedDelete = async function (args, res, next) {
+  let promises = [];
+  for (const record of records) {
+    // does this record exit in any other collection?
+    const collectionCount = await collectionDB.countDocuments({ _schemaName: RECORD_TYPE.CollectionBCMI._schemaName, records: { $elemMatch: { $eq: new ObjectID(record) } } });
```
trying to update an newly created and edited collection (note my records had no attached documents) the api crashes due to TypeError: collectionDB.countDocuments is not a function

danieltruong replied:
c4001780a78be7845c544c94e828bec4fb8b1900

mtCarto commented on angular/projects/admin-nrpti/src/app/mines/mines-collections-add-edit/mines-collections-add-edit.component.ts:
```diff
@@ -405,34 +378,30 @@ export class MinesCollectionsAddEditComponent implements OnInit, OnDestroy {
+      if (!res || res.ok !== 1) {
+        alert('Failed to create collection.');
+      } else {
+        const createdCollection = res.ops[0];
```
creating collection and saving (with no added records) gets stuck on the spinner and throws res.ops is undefined
but the collection does get created

mtCarto commented:
Creating a news item works, but throws an error and doesn't return to list page. 
Editing and deleting seem to be working

dhlevi commented:
Ah, It looks like it's the returns from the API. You've got something like:

```
obj = await Put.updateById(collectionId, updateObj);
queryActions.sendResponse(res, 200, obj);
```

On the put/posts, so it's returning the result from the mongoose process, but it looks like you meant to send back a different object. Likely the issues that we're seeing can be fixed by either changing the API up to use the custom object that was probably there originally, or just updating the angular side to work with the new results coming back.

Explains why they all seem to work, but throw errors on the angular side!

danieltruong commented:
Turns out I was using MongoDB V4 functions which makes sense that it works for me but not you guys. @dhlevi @mtCarto

danieltruong commented:
> Ah, It looks like it's the returns from the API. You've got something like:
> 
> ```
> obj = await Put.updateById(collectionId, updateObj);
> queryActions.sendResponse(res, 200, obj);
> ```
> 
> On the put/posts, so it's returning the result from the mongoose process, but it looks like you meant to send back a different object. Likely the issues that we're seeing can be fixed by either changing the API up to use the custom object that was probably there originally, or just updating the angular side to work with the new results coming back.
> 
> Explains why they all seem to work, but throw errors on the angular side!

I can't seem to get the same errors you guys are getting. I know I've changed what the api sends to the angular side, but I've also changed the angular side to account for this. What errors are being thrown in the front end?

mtCarto reviewed (approved):
Looks good, error free!

dhlevi reviewed (approved):
A lot of weird issues but looks like it's behaving itself now. Changes look good, collection and news API side are functioning as expected!
